pilot/pkg/networking/core/v1alpha3/cluster_builder.go

```diff
@@ -1143,15 +1142,24 @@ func (cb *ClusterBuilder) buildUpstreamClusterTLSContext(opts *buildClusterOpts,
 	return tlsContext, nil
 }
 
-func (cb *ClusterBuilder) setAutoSni(mc *MutableCluster, tls *networking.ClientTLSSettings) {
-	if mc != nil && features.EnableAutoSni && len(tls.Sni) == 0 {
-		if mc.httpProtocolOptions == nil {
-			mc.httpProtocolOptions = &http.HttpProtocolOptions{}
-		}
-		if mc.httpProtocolOptions.UpstreamHttpProtocolOptions == nil {
-			mc.httpProtocolOptions.UpstreamHttpProtocolOptions = &core.UpstreamHttpProtocolOptions{}
+// Set auto_sni if sni field is not explicitly set and if EnableAutoSni feature flag is enabled.
+// Set auto_san_validation if VerifyCertAtClient feature flag is enabled
+func (cb *ClusterBuilder) setAutoSniAndAutoSanValidation(mc *MutableCluster, tls *networking.ClientTLSSettings) {
+	if mc != nil && features.EnableAutoSni {
+		if len(tls.Sni) == 0 || (features.VerifyCertAtClient && len(tls.SubjectAltNames) == 0) {
```

Comment:
Why are we coupling `AutoSanValidation` with `VerifyCertAtClient`? both seems unrelated to me. 
IMO, for setting AutoSanValidation, we should just look for EnableAutoSni(may be rename it to EnableAutoSniAndSANValidation) and `len(tls.SubjectAltNames) == 0)`.

Reply:
consider the case, mode MUTUAL with user configured ca certificates. In this case VerifyCertAtClient could be false because client side does not need system ca for server cert verification, but AutoSanValidation can still be desired for auto-sni

Reply:
SAN validation doesn't even work without VerifyCertAtClient unless you are extremely careful as far as I understand. 99% of users would end up not actually validating the SAN.

VerifyCertAtClient=false is a horrible default only for backwards compatibility

Reply:
>SAN validation doesn't even work without VerifyCertAtClient unless you are extremely careful 

careful means configuring user ca certs in the DR correctly? 

>VerifyCertAtClient=false is a horrible default only for backwards compatibility

yeah, enabling it by default will mean SAN is validated using system ca certs if user has not provided ca certs and that seems better ok.

If we want to go ahead with relying on VerifyCertAtClient for AutoSanValidation, does it make sense to add validation logic to verify VerifyCertAtClient is set to true if EnableAutoSni is set to true?

Reply:
@vikaschoudhary16 : We wanted to keep EnableAutoSni separate, as that is not tightly coupled with any validation. That way it gives flexibility to use only auto_sni part. However AutoSanValidation is turned on, only if VerifyCertAtClient is enabled, and that looks better?